src/mlpack/core/util/functiontype_method_forms.hpp

```diff
@@ -0,0 +1,227 @@
+/**
+ * @file functiontype_form.hpp
+ * @author Shikhar Bhardwaj
+ *
+ * This file contains the definitions of the method forms required by the
+ * FunctionType API used by the optimizers. These method forms can be used to
+ * check the compliance of a user provided FunctionType with the required
+ * interface from the optimizer at compile time.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_CORE_FUNCTIONTYPE_METHOD_FORMS
+#define MLPACK_CORE_FUNCTIONTYPE_METHOD_FORMS
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/sfinae_utility.hpp>
+
+namespace mlpack {
+namespace static_checks {
+/*
+ * Definitions of method forms and checks for the DecomposableFunctionType API.
+ */
+HAS_METHOD_FORM(NumFunctions, HasNumFunctions);
+
+template <typename Class, typename...Ts>
+using NumFunctionsFormConst = size_t(Class::*)() const;
+
+template <typename Class, typename...Ts>
+using NumFunctionsFormNonConst = size_t(Class::*)();
+
+template<typename FunctionType>
+struct CheckNumFunctions
+{
+  const static bool value =
+    HasNumFunctions<FunctionType, NumFunctionsFormConst>::value ||
+    HasNumFunctions<FunctionType, NumFunctionsFormNonConst>::value;
+};
+
+HAS_METHOD_FORM(Evaluate, HasDecomposableEvaluate);
+
+template <typename Class, typename...Ts>
+using DecomposableEvaluateFormConst = double(Class::*)(const arma::mat&, const
+    size_t) const;
+
+template <typename Class, typename...Ts>
+using DecomposableEvaluateFormNonConst = double(Class::*)(const arma::mat&,
+    const size_t);
+
+template<typename FunctionType>
+struct CheckDecomposableEvaluate
+{
+  const static bool value =
+    HasDecomposableEvaluate<FunctionType, DecomposableEvaluateFormConst>::value
```

Comment:
See my comment for ``CheckNumFunctions``. Perhaps, that way we can prettify the expression here to improve the readability.